File: backend/solver/pulse/solver/variants/nlhe_subgame.py
"""NLHE postflop subgame solver.

Builds a game tree for No-Limit Hold'em postflop play with configurable:
- Board cards (flop/turn/river)
- Player ranges
- Bet sizes (as fractions of pot)
- Pot and stack geometry

This is the production use case — solves postflop spots given preflop ranges.
"""

from __future__ import annotations

from itertools import combinations
from typing import Sequence

from ..cards import (
    NUM_CARDS, card_from_str, card_rank, card_suit, card_to_str,
    cards_to_str, parse_range,
)
from ..evaluator import evaluate_hand
from ..game_tree import (
    Action, ActionInfo, ActionNode, ChanceNode, GameNode,
    TerminalNode, TreeBuilder,
)

DEFAULT_BET_SIZES = [0.33, 0.67, 1.0]  # fractions of pot
DEFAULT_RAISE_SIZES = [0.5, 1.0]       # fractions of pot after bet
MAX_RAISES = 3
MAX_RUNOUT_CARDS = 6  # limit runout enumeration per street for tractability


class NLHESubgameBuilder(TreeBuilder):
    """Builds an NLHE postflop subgame tree."""

    def __init__(self, board: str = "Qs9d2c",
                 range_p0: str = "AA,KK,QQ,JJ,AKs,AKo",
                 range_p1: str = "AA,KK,QQ,JJ,AKs,AKo",
                 pot: float = 100.0,
                 stack: float = 200.0,
                 bet_sizes: list[float] | None = None,
                 **kwargs):
        self.board_str = board
        self.board_cards = self._parse_board(board)
        self.range_p0 = parse_range(range_p0)
        self.range_p1 = parse_range(range_p1)
        self.pot = pot
        self.stack = stack  # effective stack remaining
        self.bet_sizes = bet_sizes or DEFAULT_BET_SIZES

        # Filter ranges: remove combos that conflict with board
        board_set = set(self.board_cards)
        self.range_p0 = [(c1, c2) for c1, c2 in self.range_p0
                         if c1 not in board_set and c2 not in board_set]
        self.range_p1 = [(c1, c2) for c1, c2 in self.range_p1
                         if c1 not in board_set and c2 not in board_set]

        # Determine street from board length
        self.street = {3: "flop", 4: "turn", 5: "river"}[len(self.board_cards)]

    def _parse_board(self, board: str) -> list[int]:
        """Parse board string like 'Qs9d2c' into card ints."""
        if len(board) % 2 != 0:
            raise ValueError(f"Invalid board string: {board}")
        cards = []
        for i in range(0, len(board), 2):
            cards.append(card_from_str(board[i:i + 2]))
        if len(cards) not in (3, 4, 5):
            raise ValueError(f"Board must be 3-5 cards, got {len(cards)}")
        return cards

    def build_tree(self) -> GameNode:
        """Build the subgame tree.

        Root is a chance node enumerating all combos of (p0_hand, p1_hand)
        that don't conflict with each other or the board.
        """
        root = ChanceNode(depth=0)

        # Enumerate all valid deal-outs
        valid_deals = []
        for h0 in self.range_p0:
            for h1 in self.range_p1:
                # No card overlap
                if h0[0] in h1 or h0[1] in h1 or h1[0] in h0 or h1[1] in h0:
                    continue
                valid_deals.append((h0, h1))

        if not valid_deals:
            raise ValueError("No valid deals — ranges may conflict with board")

        prob = 1.0 / len(valid_deals)

        for (c0a, c0b), (c1a, c1b) in valid_deals:
            label = f"{card_to_str(c0a)}{card_to_str(c0b)}v{card_to_str(c1a)}{card_to_str(c1b)}"
            child = self._build_betting(
                p0_hand=(c0a, c0b), p1_hand=(c1a, c1b),
                board=list(self.board_cards),
                pot=self.pot, stacks=(self.stack, self.stack),
                history="", num_raises=0, depth=1,
            )
            child.parent = root
            root.outcomes[label] = (prob, child)

        return root

    def _build_betting(self, p0_hand: tuple[int, int], p1_hand: tuple[int, int],
                       board: list[int], pot: float,
                       stacks: tuple[float, float],
                       history: str, num_raises: int, depth: int) -> GameNode:
        """Build a betting subtree."""
        # Determine player from action count in current street
        current_street_actions = history.split("/")[-1] if "/" in history else history
        player = len(current_street_actions) % 2

        player_hand = p0_hand if player == 0 else p1_hand
        info_key = self.get_info_set_key(
            player, list(player_hand), history, board
        )

        # Check if round is complete
        if self._is_round_complete(history):
            if history.endswith("f"):
                # Fold — previous player wins
                winner = player  # the player who DIDN'T fold
                payoff = pot / 2
                payoffs = (payoff, -payoff) if winner == 0 else (-payoff, payoff)
                return TerminalNode(depth=depth, payoffs=payoffs, pot=pot)

            # Check-check or bet-call: deal next street or showdown
            if len(board) < 5:
                return self._deal_street(
                    p0_hand, p1_hand, board, pot, stacks, history, depth
                )
            else:
                return self._showdown(p0_hand, p1_hand, board, pot, depth)

        # Build actions
        actions = self._get_actions(history, pot, stacks[player], num_raises)

        node = ActionNode(
            depth=depth,
            player=player,
            actions=actions,
            info_set_key=info_key,
            pot=pot,
            stacks=stacks,
        )

        for action in actions:
            new_stacks = list(stacks)
            if action.action == Action.FOLD:
                new_hist = history + "f"
                new_pot = pot
            elif action.action == Action.CHECK:
                new_hist = history + "x"
                new_pot = pot
            elif action.action in (Action.BET, Action.RAISE, Action.ALL_IN):
                amt = min(action.amount, stacks[player])
                new_hist = history + f"b{int(amt)}"
                new_pot = pot + amt
                new_stacks[player] -= amt
                num_raises_new = num_raises + 1
            elif action.action == Action.CALL:
                amt = action.amount
                new_hist = history + "c"
                new_pot = pot + amt
                new_stacks[player] -= amt
                num_raises_new = num_raises
            else:
                new_hist = history + "?"
                new_pot = pot

            nr = num_raises + 1 if action.action in (Action.BET, Action.RAISE, Action.ALL_IN) else num_raises

            child = self._build_betting(
                p0_hand, p1_hand, board, new_pot,
                tuple(new_stacks), new_hist, nr, depth + 1,
            )
            child.parent = node
            node.children[action.label] = child

        return node

    def _is_round_complete(self, history: str) -> bool:
        """Check if current betting round is complete."""
        current = history.split("/")[-1] if "/" in history else history
        if len(current) < 2:
            return False
        # Ends on: xx (check-check), ...c (call after bet), ...f (fold)
        if current.endswith("f"):
            return True
        if current == "xx":
            return True
        # bet followed by call
        if current.endswith("c") and len(current) >= 2:
            return True
        return False

    def _get_actions(self, history: str, pot: float, player_stack: float,
                     num_raises: int) -> list[ActionInfo]:
        """Get available actions."""
        current = history.split("/")[-1] if "/" in history else history
        facing_bet = False
        bet_amount = 0.0

        # Parse last action to detect if facing a bet
        for i in range(len(current) - 1, -1, -1):
            if current[i] == 'b':
                facing_bet = True
                # Extract amount
                j = i + 1
                while j < len(current) and current[j].isdigit():
                    j += 1
                if j > i + 1:
                    bet_amount = float(current[i + 1:j])
                break
            elif current[i] in ('x', 'c', 'f'):
                break

        if facing_bet:
            actions = [
                ActionInfo(Action.FOLD, label="fold"),
                ActionInfo(Action.CALL, amount=bet_amount, label="call"),
            ]
            # Raise options
            if num_raises < MAX_RAISES and player_stack > bet_amount:
                for frac in DEFAULT_RAISE_SIZES:
                    raise_amt = bet_amount + pot * frac
                    raise_amt = min(raise_amt, player_stack)
                    label = f"raise_{int(raise_amt)}"
                    actions.append(ActionInfo(Action.RAISE, amount=raise_amt, label=label))
                # All-in
                if player_stack > raise_amt:
                    actions.append(ActionInfo(Action.ALL_IN, amount=player_stack,
                                             label=f"allin_{int(player_stack)}"))
        else:
            actions = [ActionInfo(Action.CHECK, label="check")]
            for frac in self.bet_sizes:
                bet_amt = pot * frac
                bet_amt = min(bet_amt, player_stack)
                if bet_amt > 0:
                    label = f"bet_{int(bet_amt)}"
                    actions.append(ActionInfo(Action.BET, amount=bet_amt, label=label))
            # All-in
            if player_stack > 0:
                max_bet = max((pot * f for f in self.bet_sizes), default=0)
                if player_stack > max_bet:
                    actions.append(ActionInfo(Action.ALL_IN, amount=player_stack,
                                             label=f"allin_{int(player_stack)}"))

        return actions

    def _deal_street(self, p0_hand: tuple[int, int], p1_hand: tuple[int, int],
                     board: list[int], pot: float,
                     stacks: tuple[float, float],
                     history: str, depth: int) -> GameNode:
        """Deal the next street card as a chance node.

        To keep the tree tractable, limits to MAX_RUNOUT_CARDS representative
        cards per street (spread evenly across ranks).
        """
        used = set(board) | set(p0_hand) | set(p1_hand)
        remaining = [c for c in range(NUM_CARDS) if c not in used]

        if not remaining:
            return self._showdown(p0_hand, p1_hand, board, pot, depth)

        # Subsample for tractability
        if len(remaining) > MAX_RUNOUT_CARDS:
            step = len(remaining) / MAX_RUNOUT_CARDS
            sampled = [remaining[int(i * step)] for i in range(MAX_RUNOUT_CARDS)]
            remaining = sampled

        chance = ChanceNode(depth=depth)
        prob = 1.0 / len(remaining)

        for card_int in remaining:
            new_board = board + [card_int]
            label = card_to_str(card_int)
            new_history = history + "/" + label + "/"

            child = self._build_betting(
                p0_hand, p1_hand, new_board, pot, stacks,
                new_history, 0, depth + 1,
            )
            child.parent = chance
            chance.outcomes[label] = (prob, child)

        return chance

    def _showdown(self, p0_hand: tuple[int, int], p1_hand: tuple[int, int],
                  board: list[int], pot: float, depth: int) -> TerminalNode:
        """Evaluate hands and determine winner."""
        p0_cards = list(p0_hand) + board
        p1_cards = list(p1_hand) + board

        p0_rank = evaluate_hand(p0_cards)
        p1_rank = evaluate_hand(p1_cards)

        if p0_rank > p1_rank:
            payoff = pot / 2
            payoffs = (payoff, -payoff)
        elif p1_rank > p0_rank:
            payoff = pot / 2
            payoffs = (-payoff, payoff)
        else:
            payoffs = (0.0, 0.0)

        return TerminalNode(depth=depth, payoffs=payoffs, pot=pot, is_showdown=True)

    def get_info_set_key(self, player: int, cards: Sequence[int],
                         history: str, board: Sequence[int] = ()) -> str:
        """Info set key: 'AhKh|Qs9d2c|check:bet_50' format."""
        hand_str = cards_to_str(sorted(cards, reverse=True))
        board_str = cards_to_str(board) if board else ""
        if board_str:
            return f"{hand_str}|{board_str}|{history}"
        return f"{hand_str}|{history}"

File: lib/solver/variants/nlhe-subgame.ts
/**
 * NLHE postflop subgame solver.
 *
 * Builds a game tree for No-Limit Hold'em postflop play with configurable:
 * - Board cards (flop/turn/river)
 * - Player ranges
 * - Bet sizes (as fractions of pot)
 * - Pot and stack geometry
 */

import {
  NUM_CARDS,
  cardFromStr,
  cardToStr,
  cardsToStr,
  parseRange,
} from "../cards";
import { evaluateHand } from "../evaluator";
import {
  Action,
  makeActionInfo,
  type ActionInfo as ActionInfoType,
  type ActionNode,
  type ChanceNode,
  type GameNode,
  type TerminalNode,
  type TreeBuilder,
} from "../types";

const DEFAULT_BET_SIZES = [0.33, 0.67, 1.0];
const DEFAULT_RAISE_SIZES = [0.5, 1.0];
const MAX_RAISES = 3;
const MAX_RUNOUT_CARDS = 6;

export class NLHESubgameBuilder implements TreeBuilder {
  boardStr: string;
  boardCards: number[];
  rangeP0: [number, number][];
  rangeP1: [number, number][];
  pot: number;
  stack: number;
  betSizes: number[];
  street: string;

  constructor(opts: {
    board?: string;
    range_p0?: string;
    range_p1?: string;
    pot?: number;
    stack?: number;
    bet_sizes?: number[];
  } = {}) {
    this.boardStr = opts.board ?? "Qs9d2c";
    this.boardCards = this._parseBoard(this.boardStr);
    this.pot = opts.pot ?? 100;
    this.stack = opts.stack ?? 200;
    this.betSizes = opts.bet_sizes ?? DEFAULT_BET_SIZES;

    const boardSet = new Set(this.boardCards);
    this.rangeP0 = parseRange(opts.range_p0 ?? "AA,KK,QQ,JJ,AKs,AKo").filter(
      ([c1, c2]) => !boardSet.has(c1) && !boardSet.has(c2),
    );
    this.rangeP1 = parseRange(opts.range_p1 ?? "AA,KK,QQ,JJ,AKs,AKo").filter(
      ([c1, c2]) => !boardSet.has(c1) && !boardSet.has(c2),
    );

    const streetMap: Record<number, string> = {
      3: "flop",
      4: "turn",
      5: "river",
    };
    this.street = streetMap[this.boardCards.length];
    if (!this.street)
      throw new Error(
        `Board must be 3-5 cards, got ${this.boardCards.length}`,
      );
  }

  private _parseBoard(board: string): number[] {
    if (board.length % 2 !== 0)
      throw new Error(`Invalid board string: ${board}`);
    const cards: number[] = [];
    for (let i = 0; i < board.length; i += 2) {
      cards.push(cardFromStr(board.slice(i, i + 2)));
    }
    if (cards.length < 3 || cards.length > 5)
      throw new Error(`Board must be 3-5 cards, got ${cards.length}`);
    return cards;
  }

  buildTree(): GameNode {
    const root: ChanceNode = {
      type: "chance",
      outcomes: new Map(),
      depth: 0,
    };

    // Enumerate all valid deals
    const validDeals: [[number, number], [number, number]][] = [];
    for (const h0 of this.rangeP0) {
      for (const h1 of this.rangeP1) {
        if (
          h0[0] === h1[0] ||
          h0[0] === h1[1] ||
          h0[1] === h1[0] ||
          h0[1] === h1[1]
        )
          continue;
        validDeals.push([h0, h1]);
      }
    }

    if (validDeals.length === 0)
      throw new Error("No valid deals — ranges may conflict with board");

    const prob = 1.0 / validDeals.length;

    for (const [[c0a, c0b], [c1a, c1b]] of validDeals) {
      const label = `${cardToStr(c0a)}${cardToStr(c0b)}v${cardToStr(c1a)}${cardToStr(c1b)}`;
      const child = this._buildBetting(
        [c0a, c0b],
        [c1a, c1b],
        [...this.boardCards],
        this.pot,
        [this.stack, this.stack],
        "",
        0,
        1,
      );
      root.outcomes.set(label, { prob, child });
    }

    return root;
  }

  private _buildBetting(
    p0Hand: [number, number],
    p1Hand: [number, number],
    board: number[],
    pot: number,
    stacks: [number, number],
    history: string,
    numRaises: number,
    depth: number,
  ): GameNode {
    const slashIdx = history.lastIndexOf("/");
    const currentStreetActions =
      slashIdx >= 0 ? history.slice(slashIdx + 1) : history;
    // Count action tokens (x, f, c, or bNNN)
    let actionCount = 0;
    for (let i = 0; i < currentStreetActions.length; i++) {
      const ch = currentStreetActions[i];
      if (ch === "x" || ch === "f" || ch === "c") {
        actionCount++;
      } else if (ch === "b") {
        actionCount++;
        // Skip digits
        i++;
        while (
          i < currentStreetActions.length &&
          currentStreetActions[i] >= "0" &&
          currentStreetActions[i] <= "9"
        )
          i++;
        i--; // loop increment
      }
    }
    const player = actionCount % 2;

    const playerHand = player === 0 ? p0Hand : p1Hand;
    const infoKey = this.getInfoSetKey(
      player,
      [...playerHand],
      history,
      board,
    );

    // Check if round is complete
    if (this._isRoundComplete(history)) {
      if (history.endsWith("f")) {
        const winner = player;
        const payoff = pot / 2;
        const payoffs: [number, number] =
          winner === 0 ? [payoff, -payoff] : [-payoff, payoff];
        return { type: "terminal", depth, payoffs, pot, isShowdown: false };
      }

      // Check-check or bet-call
      if (board.length < 5) {
        return this._dealStreet(
          p0Hand,
          p1Hand,
          board,
          pot,
          stacks,
          history,
          depth,
        );
      } else {
        return this._showdown(p0Hand, p1Hand, board, pot, depth);
      }
    }

    const actions = this._getActions(history, pot, stacks[player], numRaises);
    const node: ActionNode = {
      type: "action",
      depth,
      player,
      actions,
      children: new Map(),
      infoSetKey: infoKey,
      pot,
      stacks,
    };

    for (const action of actions) {
      const newStacks: [number, number] = [...stacks];
      let newHist: string;
      let newPot: number;
      let nr: number;

      if (action.action === Action.FOLD) {
        newHist = history + "f";
        newPot = pot;
        nr = numRaises;
      } else if (action.action === Action.CHECK) {
        newHist = history + "x";
        newPot = pot;
        nr = numRaises;
      } else if (
        action.action === Action.BET ||
        action.action === Action.RAISE ||
        action.action === Action.ALL_IN
      ) {
        const amt = Math.min(action.amount, stacks[player]);
        newHist = history + `b${Math.floor(amt)}`;
        newPot = pot + amt;
        newStacks[player] -= amt;
        nr = numRaises + 1;
      } else if (action.action === Action.CALL) {
        const amt = action.amount;
        newHist = history + "c";
        newPot = pot + amt;
        newStacks[player] -= amt;
        nr = numRaises;
      } else {
        newHist = history + "?";
        newPot = pot;
        nr = numRaises;
      }

      const child = this._buildBetting(
        p0Hand,
        p1Hand,
        board,
        newPot,
        newStacks,
        newHist,
        nr,
        depth + 1,
      );
      node.children.set(action.label, child);
    }

    return node;
  }

  private _isRoundComplete(history: string): boolean {
    const slashIdx = history.lastIndexOf("/");
    const current = slashIdx >= 0 ? history.slice(slashIdx + 1) : history;
    if (current.length < 2) return false;

    if (current.endsWith("f")) return true;
    if (current === "xx") return true;
    // bet followed by call
    if (current.endsWith("c") && current.length >= 2) return true;
    return false;
  }

  private _getActions(
    history: string,
    pot: number,
    playerStack: number,
    numRaises: number,
  ): ActionInfoType[] {
    const slashIdx = history.lastIndexOf("/");
    const current = slashIdx >= 0 ? history.slice(slashIdx + 1) : history;
    let facingBet = false;
    let betAmount = 0;

    // Parse last action to detect if facing a bet
    for (let i = current.length - 1; i >= 0; i--) {
      if (current[i] === "b") {
        facingBet = true;
        let j = i + 1;
        while (j < current.length && current[j] >= "0" && current[j] <= "9")
          j++;
        if (j > i + 1) betAmount = parseFloat(current.slice(i + 1, j));
        break;
      } else if (
        current[i] === "x" ||
        current[i] === "c" ||
        current[i] === "f"
      ) {
        break;
      }
    }

    if (facingBet) {
      const actions: ActionInfoType[] = [
        makeActionInfo(Action.FOLD, 0, "fold"),
        makeActionInfo(Action.CALL, betAmount, "call"),
      ];
      // Raise options
      if (numRaises < MAX_RAISES && playerStack > betAmount) {
        let lastRaiseAmt = 0;
        for (const frac of DEFAULT_RAISE_SIZES) {
          let raiseAmt = betAmount + pot * frac;
          raiseAmt = Math.min(raiseAmt, playerStack);
          lastRaiseAmt = raiseAmt;
          actions.push(
            makeActionInfo(
              Action.RAISE,
              raiseAmt,
              `raise_${Math.floor(raiseAmt)}`,
            ),
          );
        }
        // All-in
        if (playerStack > lastRaiseAmt) {
          actions.push(
            makeActionInfo(
              Action.ALL_IN,
              playerStack,
              `allin_${Math.floor(playerStack)}`,
            ),
          );
        }
      }
      return actions;
    }

    const actions: ActionInfoType[] = [
      makeActionInfo(Action.CHECK, 0, "check"),
    ];
    for (const frac of this.betSizes) {
      let betAmt = pot * frac;
      betAmt = Math.min(betAmt, playerStack);
      if (betAmt > 0) {
        actions.push(
          makeActionInfo(
            Action.BET,
            betAmt,
            `bet_${Math.floor(betAmt)}`,
          ),
        );
      }
    }
    // All-in
    if (playerStack > 0) {
      const maxBet = Math.max(
        ...this.betSizes.map((f) => pot * f),
        0,
      );
      if (playerStack > maxBet) {
        actions.push(
          makeActionInfo(
            Action.ALL_IN,
            playerStack,
            `allin_${Math.floor(playerStack)}`,
          ),
        );
      }
    }
    return actions;
  }

  private _dealStreet(
    p0Hand: [number, number],
    p1Hand: [number, number],
    board: number[],
    pot: number,
    stacks: [number, number],
    history: string,
    depth: number,
  ): GameNode {
    const used = new Set([...board, ...p0Hand, ...p1Hand]);
    let remaining: number[] = [];
    for (let c = 0; c < NUM_CARDS; c++) {
      if (!used.has(c)) remaining.push(c);
    }

    if (remaining.length === 0) {
      return this._showdown(p0Hand, p1Hand, board, pot, depth);
    }

    // Subsample for tractability
    if (remaining.length > MAX_RUNOUT_CARDS) {
      const step = remaining.length / MAX_RUNOUT_CARDS;
      const sampled: number[] = [];
      for (let i = 0; i < MAX_RUNOUT_CARDS; i++) {
        sampled.push(remaining[Math.floor(i * step)]);
      }
      remaining = sampled;
    }

    const chance: ChanceNode = {
      type: "chance",
      outcomes: new Map(),
      depth,
    };
    const prob = 1.0 / remaining.length;

    for (const cardInt of remaining) {
      const newBoard = [...board, cardInt];
      const label = cardToStr(cardInt);
      const newHistory = history + "/" + label + "/";
      const child = this._buildBetting(
        p0Hand,
        p1Hand,
        newBoard,
        pot,
        stacks,
        newHistory,
        0,
        depth + 1,
      );
      chance.outcomes.set(label, { prob, child });
    }

    return chance;
  }

  private _showdown(
    p0Hand: [number, number],
    p1Hand: [number, number],
    board: number[],
    pot: number,
    depth: number,
  ): TerminalNode {
    const p0Cards = [...p0Hand, ...board];
    const p1Cards = [...p1Hand, ...board];

    const p0Rank = evaluateHand(p0Cards);
    const p1Rank = evaluateHand(p1Cards);

    let payoffs: [number, number];
    if (p0Rank > p1Rank) {
      payoffs = [pot / 2, -pot / 2];
    } else if (p1Rank > p0Rank) {
      payoffs = [-pot / 2, pot / 2];
    } else {
      payoffs = [0, 0];
    }

    return { type: "terminal", depth, payoffs, pot, isShowdown: true };
  }

  getInfoSetKey(
    player: number,
    cards: number[],
    history: string,
    board?: number[],
  ): string {
    const handStr = cardsToStr([...cards].sort((a, b) => b - a));
    const boardStr = board && board.length > 0 ? cardsToStr(board) : "";
    if (boardStr) return `${handStr}|${boardStr}|${history}`;
    return `${handStr}|${history}`;
  }
}
